File: test-fixtures/languages/kotlin/complex/DeepFlowChain.kt
package com.pipeline.entry

// DeepFlowChain.kt -- Entry point for deep cross-file flow chain.
// All type/function names prefixed with Dfc.

annotation class DfcAudited

object DfcCommandParser {
    fun tokenize(input: String): List<String> {
        val tokens = input.trim().split("\\s+".toRegex())
        return tokens
    }

    fun extractAction(tokens: List<String>): String {
        if (tokens.isEmpty()) return ""
        var action = tokens[0]
        for (i in 1 until tokens.size) {
            action = action + " " + tokens[i]
        }
        return action
    }
}

object DfcInputValidator {
    fun validateLength(input: String): String {
        var validated = input
        if (validated.length > 4096)
            validated = validated.substring(0, 4096)
        return validated
    }

    fun normalizeWhitespace(input: String): String {
        val normalized = input.trim().replace("\\s+".toRegex(), " ")
        return normalized
    }
}

data class DfcRequestContext(
    val action: String,
    val origin: String,
    val metadata: MutableMap<String, String> = mutableMapOf()
)

@DfcAudited
fun main() {
    // SOURCE: user-controlled stdin input
    val rawInput = readLine() ?: return

    val tokens = DfcCommandParser.tokenize(rawInput)
    val action = DfcCommandParser.extractAction(tokens)
    val validated = DfcInputValidator.validateLength(action)
    val normalized = DfcInputValidator.normalizeWhitespace(validated)

    val ctx = DfcRequestContext(normalized, "cli")

    // Cross to helpers
    val registry = DfcPipelineRegistry()
    registry.register("default", ctx)
    val registeredCtx = registry.lookup("default")

    val prefixed = DfcTransformStage.applyPrefix(registeredCtx!!)
    val wrapped = DfcTransformStage.wrapInQuotes(prefixed)

    val assembler = DfcCommandAssembler()
    assembler.setPayload(wrapped)
    assembler.setTarget("/bin/sh")
    val assembled = assembler.assemble()

    val formatted = DfcOutputFormatter.formatForShell(assembled)

    // Cross to sink
    val executor = DfcCommandExecutor()
    val prepared = executor.prepare(formatted)
    val finalCommand = executor.constructCommand(prepared)

    val execCtx = DfcExecutionContext()
    execCtx.setCommand(finalCommand)
    val cmdToRun = execCtx.getCommand()

    DfcAuditLogger.logPreExecution(cmdToRun)

    val launcher = DfcProcessLauncher()
    launcher.createProcess(cmdToRun)
    launcher.configureEnvironment()
    launcher.launch()  // SINK: Runtime.exec
}

File: test-fixtures/languages/csharp/complex/DeepFlowChain.cs
using System;
using System.Collections.Generic;

// DeepFlowChain.cs -- Entry point for deep cross-file flow chain.
// All type/function names prefixed with Dfc.

namespace Pipeline.Entry
{
    public class DfcAuditedAttribute : Attribute { }

    public class DfcCommandParser
    {
        public static string[] Tokenize(string input)
        {
            string[] tokens = input.Trim().Split(new[] { ' ', '\t' },
                StringSplitOptions.RemoveEmptyEntries);
            return tokens;
        }

        public static string ExtractAction(string[] tokens)
        {
            if (tokens.Length == 0) return "";
            string action = tokens[0];
            for (int i = 1; i < tokens.Length; i++)
            {
                action = action + " " + tokens[i];
            }
            return action;
        }
    }

    public class DfcInputValidator
    {
        public static string ValidateLength(string input)
        {
            string validated = input;
            if (validated.Length > 4096)
                validated = validated.Substring(0, 4096);
            return validated;
        }

        public static string NormalizeWhitespace(string input)
        {
            string[] parts = input.Split(new[] { ' ' },
                StringSplitOptions.RemoveEmptyEntries);
            string normalized = string.Join(" ", parts);
            return normalized;
        }
    }

    public class DfcRequestContext
    {
        public string Action { get; set; }
        public string Origin { get; set; }

        public DfcRequestContext(string action, string origin)
        {
            Action = action;
            Origin = origin;
        }
    }

    public class DeepFlowChain
    {
        [DfcAudited]
        public static void Main(string[] args)
        {
            // SOURCE: user-controlled stdin input
            string rawInput = Console.ReadLine();

            string[] tokens = DfcCommandParser.Tokenize(rawInput);
            string action = DfcCommandParser.ExtractAction(tokens);
            string validated = DfcInputValidator.ValidateLength(action);
            string normalized = DfcInputValidator.NormalizeWhitespace(validated);

            var ctx = new DfcRequestContext(normalized, "cli");

            // Cross to helpers
            var registry = new DfcPipelineRegistry();
            registry.Register("default", ctx);
            var registeredCtx = registry.Lookup("default");

            string prefixed = DfcTransformStage.ApplyPrefix(registeredCtx);
            string wrapped = DfcTransformStage.WrapInQuotes(prefixed);

            var assembler = new DfcCommandAssembler();
            assembler.SetPayload(wrapped);
            assembler.SetTarget("cmd.exe");
            string assembled = assembler.Assemble();

            string formatted = DfcOutputFormatter.FormatForShell(assembled);

            // Cross to sink
            var executor = new DfcCommandExecutor();
            string prepared = executor.Prepare(formatted);
            string finalCommand = executor.ConstructCommand(prepared);

            var execCtx = new DfcExecutionContext();
            execCtx.SetCommand(finalCommand);
            string cmdToRun = execCtx.GetCommand();

            var launcher = new DfcProcessLauncher();
            launcher.CreateProcess(cmdToRun);
            launcher.ConfigureEnvironment();
            launcher.Launch();  // SINK: Process.Start
        }
    }
}
